src/GafferCycles/IECoreCyclesPreview/Renderer.cpp

```diff
@@ -3649,6 +3649,8 @@ class CyclesRenderer final : public IECoreScenePreview::Renderer
 				m_scene->film->copy_value(socketType, m_film, *m_film.type->find_input( socketType.name ) );
 			}
 
+			m_scene->background->set_shader( m_scene->default_background );
```

Comment:
> Without the background pointer fix one on resets, all I need to do is make a sphere and view it through the viewport for a 100% reliable crash

I can't reproduce that at all here - is this another one that only shows up when Cycles itself is built in debug mode? What I _can_ reproduce here is that the sphere doesn't show up on the first refresh, and only shows when I move the camera for the first time. And somehow _that_ is fixed by this line.

I would still love test coverage for this, but in the absence of that, let me use the opportunity to learn more. You've mentioned before that you removed support for some interactive edits because they required a session reset of some sort, and that was unreliable. Was that _different_ to the reset that we are doing here?

Reply:
> > Without the background pointer fix one on resets, all I need to do is make a sphere and view it through the viewport for a 100% reliable crash
> 
> I can't reproduce that at all here - is this another one that only shows up when Cycles itself is built in debug mode? What I _can_ reproduce here is that the sphere doesn't show up on the first refresh, and only shows when I move the camera for the first time. And somehow _that_ is fixed by this line.
> 
Hmm I've been running cycles 3.3.0 here locally which might be a factor here as well, a stale pointer of the background shader could be a more benign side-effect of the sphere not showing up in 3.2.0 but in 3.3.0 it is crashing. My reset code above where it transfers the socket data I'm thinking if the data is not a basic type but rather a pointer to a shader in this case, then it could be a vector for crashes. [This](https://github.com/blender/cycles/blob/master/src/scene/background.cpp#L89) is the line where it crashes.

> I would still love test coverage for this, but in the absence of that, let me use the opportunity to learn more. You've mentioned before that you removed support for some interactive edits because they required a session reset of some sort, and that was unreliable. Was that _different_ to the reset that we are doing here?

Yeah the hard reset code I'd like to phase out but there is still that initial start where before rendering commences and any of the SceneParams and SessionParams differ to the defaults, the reset() code needs to still function here eg. if the device is GPU and not CPU, SVM over OSL, etc. One alternative could be that we defer the creation to render() so that all options are set and there doesn't need to be a reset in the first place. One better option that I'd like to see in the renderer interface is a way to do a hard reset similar to when a failed attribute issues another object rebuild, a global option could re-issue a full reset of the renderer. 

I can't see that texture limit option working for the viewport at all because of this which is a shame...

Reply:
> One alternative could be that we defer the creation to render() so that all options are set and there doesn't need to be a reset in the first place.

That sounds good to me if we can't make the reset reliable. It would make everything easier to reason about knowing that the state was created once and was then immutable...